src/engine/renderer/tr_shader.cpp

```diff
@@ -5264,6 +5311,140 @@ static void CollapseStages()
 			}
 		}
 	}
+
+	GroupActiveStages();
+}
+
+// Preselect the renderers to be used to render the stages.
+static void SetStagesRenderers()
+{
+	struct stageRenderers_t {
+		stageRenderer_t generic;
+		bool doShadowFill;
+		bool doForwardLighting;
+	};
+
+	typedef std::unordered_map<stageType_t, stageRenderers_t> stageTypeRenderersMap_t;
+
+	static const stageTypeRenderersMap_t stageTypeRenderers = {
+		{ stageType_t::ST_COLORMAP, {
+			&Render_generic,
+			true,
+			false,
+		} },
+		{ stageType_t::ST_STYLELIGHTMAP, {
+			&Render_generic3D,
+			true,
+			false,
+		} },
+		{ stageType_t::ST_STYLECOLORMAP, {
+			&Render_generic3D,
+			true,
+			false,
+		} },
+		{ stageType_t::ST_LIGHTMAP, {
+			&Render_lightMapping,
+			true,
+			true,
+		} },
+		{ stageType_t::ST_DIFFUSEMAP, {
+			&Render_lightMapping,
+			true,
+			true,
+		} },
+		{ stageType_t::ST_COLLAPSE_COLORMAP, {
+			&Render_lightMapping,
+			true,
+			false,
+		} },
+		{ stageType_t::ST_COLLAPSE_DIFFUSEMAP, {
+			&Render_lightMapping,
+			true,
+			true,
+		} },
+		{ stageType_t::ST_REFLECTIONMAP, {
+			&Render_reflection_CB,
+			false,
+			false,
+		} },
+		{ stageType_t::ST_COLLAPSE_REFLECTIONMAP, {
+			&Render_reflection_CB,
+			false,
+			false,
+		} },
+		{ stageType_t::ST_SKYBOXMAP, {
+			&Render_skybox,
+			false,
+			false,
+		} },
+		{ stageType_t::ST_SCREENMAP, {
+			&Render_screen,
+			false,
+			false,
+		} },
+		{ stageType_t::ST_PORTALMAP, {
+			&Render_portal,
+			false,
+			false,
+		} },
+		{ stageType_t::ST_HEATHAZEMAP, {
+			&Render_heatHaze,
+			false,
+			false,
+		} },
+		{ stageType_t::ST_LIQUIDMAP, {
+			&Render_liquid,
+			false,
+			false,
+		} },
+		{ stageType_t::ST_ATTENUATIONMAP_XY, {
+			&Render_NOP,
+			false,
+			false,
+		} },
+		{ stageType_t::ST_ATTENUATIONMAP_Z, {
+			&Render_NOP,
+			false,
+			false,
+		} },
+	};
+
+	for ( int s = 0; s < shader.numStages; s++ )
+	{
+		shaderStage_t *stage = &stages[ s ];
+
+		stage->genericRenderer = &Render_NOP;
+		stage->doShadowFill = false;
+		stage->doForwardLighting = false;
+
```

Comment:
How about ASSERTing the `count` to avoid looking up the key twice in release mode?

Reply:
This code is not that critical for performance. You said [here](https://github.com/DaemonEngine/Daemon/pull/1076#issuecomment-2053715689) that calling `.at()` on a `const unordered_map` with a missing key would crash, and if it's true it would even crash in release code, even when `ASSERT` is disabled.

If we care about performance we can do a switch case.

Reply:
I did a switch case, after all my rewrites, it's now a good solution again.